### PR TITLE
Adding my comments to Intersection Types documentation

### PR DESCRIPTION
**Pull request by radeusgd**
### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

### CODE REVIEW

File: docs/types/intersection-types.md [hunk 1]
```diff
@@ -42,17 +42,61 @@ Complex.from (that:Float) = Complex.Num that 0
...
+> ```
+> f a:Float = a
+> g a = a:Float
+> ```
+> These two functions, while very similar, will have different behaviour when 
+> passed a value like the value `c` above. The function `f` will fail with 
+> a type error, because the visible type of `c` is just `Complex` (assuming 
+> the conversion to `Float` is not available in the current scope). 
+> However, the function `g` will accept the same value and return it as 
+> a `Float` value, based on the 'hidden' part of its type.
```
Comment (radeusgd) [inline docs/types/intersection-types.md:99]: I want to make it clear that while the syntax is very similar, the checked argument and type-asserting expressions are actually different and behave differently in regard to the 'hidden' types.

File: docs/types/intersection-types.md [hunk 2]
```diff
@@ -42,17 +42,61 @@ Complex.from (that:Float) = Complex.Num that 0
...
+> [!NOTE]
+> Note that if a `Float.from (that:Complex)` conversion were available in the scope, 
+> any `Complex` instance would be convertible to `Float` regardless of how it was constructed. 
+> To ensure that such mix-ins are only available on values that opt-in to being 
+> an intersection type (like in the `Complex` example above where we include 
+> the `Float` mix-in only if `self.im == 0`), we need to ensure that the conversion 
+> used to create the intersection type is not available in the default 
+> conversion resolution scope. Thus it cannot be defined in the same module 
+> as `Complex` or `Float` types, but instead it should be defined in a separate 
+> module that is only imported in the place that will be constructing the multi-values.
```
Comment (JaroslavTulach) [inline docs/types/intersection-types.md:54]: That's correct and exactly as [prototyped in the tests](https://github.com/enso-org/enso/commit/3d8a0e1b90b20cfdfe5da8d2d3950f644a4b45b8#diff-c6ef852899778b52ce6a11ebf9564d102c273021b212a4848b7678e120776287R23).

File: docs/types/intersection-types.md [hunk 3]
```diff
@@ -42,17 +42,61 @@ Complex.from (that:Float) = Complex.Num that 0
...
+method calls will also only accept the value if it satisfies the type it 
+_has been cast to_. Any additional remaining types that it had before the cast 
+become hidden and can only be brought back through an _explicit_ cast.
+
+To perform an explicit cast that can uncover the 'hidden' part of a type, one can either:
+- write a type asserting expression - e.g. `f = c:Float`,
+- or inspect the types in a `case` expression, e.g.
+  ```
+  case c of
+      f : Float -> f.sqrt
```
Comment (JaroslavTulach) [inline docs/types/intersection-types.md:83]: Does not work. Reported as
- #11828

**Review by JaroslavTulach — CHANGES_REQUESTED** [review]
Accepted most of your changes in https://github.com/enso-org/enso/pull/11600/commits/d1cbc8d7d6f04f811a18e51afd0331b3853b2289

**Comment by JaroslavTulach** [comment]
Accepted as https://github.com/enso-org/enso/pull/11600/commits/d1cbc8d7d6f04f811a18e51afd0331b3853b2289

**Pull request closed without merging**